src/server/operators/process.rs

```diff
@@ -0,0 +1,212 @@
+use crate::server::operators::select::*;
+
+// Wrapper for DNF converter; converts a Select to DNF.
+pub fn dnf(s: Select) -> Select {
+    let new_cond = dnf_helper(s.predicate.condition);
+    let new_pred = Predicate { name: s.predicate.name, condition: new_cond };
+    Select { name: s.name, predicate: new_pred }
+}
+
+// Converts a predicate to DNF.
+fn dnf_helper(f: Conditions) -> Conditions {
+    // If the subtree has no Ors, terminate.
+    if is_all_and(f.clone()) {
+        f
+    }
+    
+    // Otherwise, we need to do pushdown.
+    else {
+        match f {
+            // If this is a Leaf, return.
+            Conditions::Leaf(_) => f,
+
+            // If this is an Or, then evaluate the children; we can safely skip this.
+            Conditions::Or(l, r) => Conditions::Or(Box::new(dnf_helper(*l)), Box::new(dnf_helper(*r))),
+
+            // If this is an And...
+            Conditions::And(l, r) => {
+                // If either left or right is an Or, then we distribute.
+                if is_or((*l).clone()) {
+                    let (lp, rp) = pushdown_disjunction(*r, *l);
+                    Conditions::Or(Box::new(dnf_helper(lp)), Box::new(dnf_helper(rp)))
+                } else if is_or((*r).clone()) {
+                    let (lp, rp) = pushdown_disjunction(*l, *r);
+                    Conditions::Or(Box::new(dnf_helper(lp)), Box::new(dnf_helper(rp)))
+                }
+                
+                // If neither child of the And is an Or, then we need to revisit later.
+                else {
+                    dnf_helper(Conditions::And(Box::new(dnf_helper(*l)), Box::new(dnf_helper(*r))))
+                }
+            }
+        }
+    }
+}
+
+// Generates the left and right after distributing x over or.
+fn pushdown_disjunction(x: Conditions, or: Conditions) -> (Conditions, Conditions) {
+    match or {
+        Conditions::Or(orl, orr) => {
+            (Conditions::And(orl, Box::new(x.clone())), Conditions::And(orr, Box::new(x.clone())))
```

Comment:
Not sure how expensive the recursive operation is, but we can save duplicate computation by calling `dnf_helper(*x)`, then cloning the result.

We can also save some checks by calling `dnf_helper(*orl)` and `dnf_helper(*orr)` directly here, but this is mostly insignificant.